### PR TITLE
Make GrMhd and GhMhd executables use runtime initial data.  

### PR DESCRIPTION
**Pull request by isaaclegred**
## Proposed changes

This PR makes the `ValenciaDivClean` and `GhValenciaDivClean` executables use runtime initial data.  This reduces the number of executables that need to be built to fully test the code, and simplifies logic in certain places.  In addition, a thermodynamic dimension-3 EoS is standardized as the EoS used in GrMhd and GhMhd, since the EoS type can no longer be simply extracted from the initial data at compile time.  

### Upgrade instructions

All Initial Data will now be identified by `InitialData` in the input file, rather than `AnalyticSolution`, `AnalyticData`, or `InitialData`.

The EoS must be explicitly identified in the input file for all evolutions.  To achieve "legacy" behavior, of using the EoS associated with the Initial data, specify 
`EquationOfState: FromInitialData`
This will extract the EoS from the initial data and promote it to a 3-d EoS for evolution.  

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # Distributed under the MIT License.
 # See LICENSE.txt for details.
 
+
+
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt]: Remove empty lines :)

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt [hunk 2]
```diff
@@ -66,14 +67,14 @@ endfunction()
 
 add_ghmhd_executable(
   ""
-  evolution::NumericInitialData
-  false
+  "false"
+  "${LIBS_TO_LINK}"
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt]: You don't need this line

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt [hunk 3]
```diff
@@ -66,14 +67,14 @@ endfunction()
...
   )
 
 add_ghmhd_executable(
   "Bns"
-  evolution::NumericInitialData
-  true
+   "true"
+  "${LIBS_TO_LINK_WITH_CONTROL_SYSTEM}"
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt]: Also don't need this line

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt [hunk 4]
```diff
@@ -86,43 +87,11 @@ function(add_ghmhd_with_horizon_executable SUFFIX INITIAL_DATA)
   target_compile_definitions(
     ${EXECUTABLE}
     PRIVATE
-    INITIAL_DATA=${INITIAL_DATA}
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/CMakeLists.txt:89]: Delete the whole function call which is now empty

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 5]
```diff
@@ -367,39 +349,26 @@ template <typename EvolutionMetavarsDerived, bool UseDgSubcell,
...
-         Parallel::Phase::Exit}};
-  } else {
-    return std::array<Parallel::Phase, 6>{
-        {Parallel::Phase::Initialization,
-         Parallel::Phase::InitializeInitialDataDependentQuantities,
-         Parallel::Phase::InitializeTimeStepperHistory,
-         Parallel::Phase::Register, Parallel::Phase::Evolve,
-         Parallel::Phase::Exit}};
-  }
+  return std::array<Parallel::Phase, 8>{
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp]: You can inline this now where it's used below

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 6]
```diff
@@ -420,37 +389,11 @@ struct GhValenciaDivCleanTemplateBase<
...
-  using analytic_solutions_and_data_to_register =
-      tmpl::append<gh::solutions_including_matter<3>,
-                   grmhd::ValenciaDivClean::InitialData::initial_data_list>;
-
-  using initial_data_tag =
-      tmpl::conditional_t<is_analytic_solution_v<initial_data>,
-                          Tags::AnalyticSolution<initial_data>,
-                          Tags::AnalyticData<initial_data>>;
+  using initial_data_list =
+      tmpl::append<ghmhd::GhValenciaDivClean::InitialData::initial_data_list>;
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp]: Remove `tmpl::append`

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 7]
```diff
@@ -587,6 +530,7 @@ struct GhValenciaDivCleanTemplateBase<
           >,
       tmpl::conditional_t<
           use_dg_subcell,
+
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp]: Delete empty line

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 8]
```diff
@@ -657,30 +601,28 @@ struct GhValenciaDivCleanTemplateBase<
...
-      tmpl::list<::Tags::DerivCompute<
+      tmpl::list<analytic_compute, error_compute>,
+          tmpl::list<::Tags::DerivCompute<
                      typename system::variables_tag,
                      ::Events::Tags::ObserverMesh<volume_dim>,
                      ::Events::Tags::ObserverInverseJacobian<
                          volume_dim, Frame::ElementLogical, Frame::Inertial>,
                      typename system::gradient_variables>,
-                 gh::gauges::Tags::GaugeAndDerivativeCompute<volume_dim>>>;
+
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp]: Run clang-format please

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 9]
```diff
@@ -713,20 +655,7 @@ struct GhValenciaDivCleanTemplateBase<
...
-                    grmhd::GhValenciaDivClean::NumericInitialData,
-                    tmpl::conditional_t<
-                        std::is_same_v<SpecInitialData<thermodynamic_dim>,
-                                       NoSuchType>,
-                        tmpl::list<>, SpecInitialData<thermodynamic_dim>>,
-                    tmpl::conditional_t<
-                        std::is_same_v<FukaInitialData, NoSuchType>,
-                        tmpl::list<>, FukaInitialData>>>,
-                analytic_solutions_and_data_to_register>>,
+        tmpl::pair<evolution::initial_data::InitialData, initial_data_list>,
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:615]: Can SpEC and FUKA ID remain here?

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 10]
```diff
@@ -903,11 +828,9 @@ struct GhValenciaDivCleanTemplateBase<
       tmpl::conditional_t<
           use_dg_subcell,
           tmpl::list<evolution::dg::subcell::Actions::SetSubcellGrid<
-              volume_dim, system, use_numeric_initial_data>>,
-          tmpl::conditional_t<use_numeric_initial_data, tmpl::list<>,
-                              evolution::Initialization::Actions::SetVariables<
-                                  ::domain::Tags::Coordinates<
-                                      volume_dim, Frame::ElementLogical>>>>,
+              volume_dim, system, true>>,
+          evolution::Initialization::Actions::SetVariables<
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:798]: On mobile now so can't check this easily: does this still work with numeric ID? Before this action was just skipped.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt [hunk 11]
```diff
@@ -1,6 +1,5 @@
 # Distributed under the MIT License.
 # See LICENSE.txt for details.
-
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt]: Please don't add or delete empty lines in various places :D what I usually do is look over the diff in VSCode and drop changes that are unrelated to the PR.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt [hunk 12]
```diff
@@ -44,122 +43,27 @@ function(add_grmhd_executable SUFFIX INITIAL_DATA INTERPOLATION_TARGET)
...
-  ""
-  )
+  false
+)
 
+# Core-collpase SN uses parametrized deleptonization
 add_grmhd_executable(
-  MagneticFieldLoop
-  grmhd::AnalyticData::MagneticFieldLoop
+  CCSN
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt]: @pajkosmi do you have a wish for the name of your executable? Suggestions: `EvolveCcsn`, `EvolveCoreCollapse`, `EvolveSupernova`, `MikesExecutable`

**Review by nilsvu — CHANGES_REQUESTED** [review]
Reviewed on mobile, so a bit brief, but couldn't resist looking at this :) Looks good!

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt (hunk 12 above)
Comment (pajkosmi) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt]: `EvolveCoreCollapseSupernova` is my vote.  It's a tad long but the most precise.

Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt]: Convention from other executables would be `EvolveGhValenciaDivCleanCcsn`. For example, we have `EvolveGhValenciaDivCleanBns`. I'd prefer to follow conventions since that makes finding executables easier.

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp (hunk 9 above)
Comment (isaaclegred) [reply src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:615]: Hm I think this logic should probably be moved to `all_solutions.hpp`.  It looks to me like FUKA wasn't moved correctly and I can fix that

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt (hunk 12 above)
Comment (isaaclegred) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt]: Yeah I do think it's important for GhValenciaDivClean to be in there somehow, current consensus would be `EvolveGhValenciaDivCleanCoreCollapseSupernova` but that's a bit long.

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp (hunk 9 above)
Comment (isaaclegred) [reply src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:658]: I think this logic moved to `all_solutions.hpp`,  I don't think I got the FUKA part yet but the SpEC part should be there.

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt (hunk 12 above)
Comment (nilsvu) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt]: Maybe shorten to `EvolveGhmhdCcsn` (and also `EvolveGhmhdBns` etc)? That's a lot of abbreviations but it's short :)

Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/CMakeLists.txt]: I think `EvolveGhValenciaDivCleanCcsn` is fine. Again, we have `EvolveGhValenciaDivCleanBns` already, which I think has a reasonable balance of length and description. There are other formulations of MHD that one could try to use, so I don't like using `mhd`.

**Comment by isaaclegred** [comment]
@nilsvu this is ready to look at again.  Let me know and I can squash/rebase on develop

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp (hunk 7 above)
Comment (nilsvu) [reply src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp]: ping

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 13]
```diff
@@ -657,30 +586,28 @@ struct GhValenciaDivCleanTemplateBase<
...
+      tmpl::list<analytic_compute, error_compute>,
+      tmpl::list<
+          ::Tags::DerivCompute<
+              typename system::variables_tag,
+              ::Events::Tags::ObserverMesh<volume_dim>,
+              ::Events::Tags::ObserverInverseJacobian<
+                  volume_dim, Frame::ElementLogical, Frame::Inertial>,
+              typename system::gradient_variables>,
+
+          tmpl::list<gh::gauges::Tags::GaugeAndDerivativeCompute<volume_dim>>>>;
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp]: Remove nested `tmpl::list` here and empty line above (there's probably a `tmpl::flatten` somewhere which is why this still compiles but can't rely on it)

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp (hunk 9 above)
Comment (nilsvu) [reply src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:615]: Ok. Please remove the SpEC/FUKA code at the beginning of this file then.

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp (hunk 10 above)
Comment (nilsvu) [reply src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp:798]: I think this conditional branch is never tested and probably doesn't work, because `use_dg_subcell` is always toggled on. Maybe add a comment that it hasn't been tested?

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp [hunk 14]
```diff
@@ -461,7 +457,7 @@ struct EvolutionMetavars<InitialData, tmpl::list<InterpolationTargetTags...>> {
 
       Actions::Label<evolution::dg::subcell::Actions::Labels::BeginSubcell>,
       Actions::MutateApply<evolution::dg::subcell::BackgroundGrVars<
-          system, EvolutionMetavars, false, false>>,
+          system, EvolutionMetavars, true, true>>,
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp]: Should the second boolean be toggled on? The parameter is `ComputeOnlyOnRollback` but I don't find any docs for that. @nilsdeppe?

File: tests/InputFiles/GrMhd/ValenciaDivClean/FishboneMoncriefDisk.yaml [hunk 15]
```diff
@@ -1,7 +1,7 @@
 # Distributed under the MIT License.
 # See LICENSE.txt for details.
 
-Executable: EvolveValenciaDivCleanFishboneMoncriefDisk
+Executable: EvolveValenciaDivClean
```
Comment (nilsvu) [inline tests/InputFiles/GrMhd/ValenciaDivClean/FishboneMoncriefDisk.yaml]: should be WithHorizon I think?

File: src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp [hunk 16]
```diff
@@ -657,30 +586,28 @@ struct GhValenciaDivCleanTemplateBase<
...
-                tmpl::list<grmhd::GhValenciaDivClean::BoundaryConditions::
-                               DirichletAnalytic,
-                           grmhd::GhValenciaDivClean::BoundaryConditions::
-                               DirichletFreeOutflow>>>,
+        tmpl::append<grmhd::GhValenciaDivClean::BoundaryConditions::
+                         standard_fd_boundary_conditions,
+                     tmpl::list<grmhd::GhValenciaDivClean::BoundaryConditions::
+                                    DirichletAnalytic,
+                                grmhd::GhValenciaDivClean::BoundaryConditions::
+                                    DirichletFreeOutflow>>,
```
Comment (nilsvu) [inline src/Evolution/Executables/GrMhd/GhValenciaDivClean/GhValenciaDivCleanBase.hpp]: you can move these into `standard_fd_boundary_conditions` (I think I moved them out of there when adding numeric ID, but now they can move back)

File: src/PointwiseFunctions/Hydro/Tags.hpp [hunk 17]
```diff
@@ -29,6 +35,20 @@ struct EquationOfState {
 };
 }  // namespace OptionTags
 
+namespace evolution::grmhd {
```
Comment (nilsvu) [inline src/PointwiseFunctions/Hydro/Tags.hpp]: remove this namespace since you're in the `hydro` namespace/directory/file here. If you did this to avoid name collisions for the `EquationOfState` tag you can either name it something slightly different or move it to the `Evolution/GRMHD` directory.

**Review by nilsvu — CHANGES_REQUESTED** [review]
squash 👍

File: src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp (hunk 14 above)
Comment (isaaclegred) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp]: Ah yeah I'll leave it false

Comment (nilsdeppe) [reply src/Evolution/Executables/GrMhd/ValenciaDivClean/EvolveValenciaDivClean.hpp]: Yes, I think leaving it as false is the right thing to do

File: src/PointwiseFunctions/Hydro/Tags.hpp (hunk 17 above)
Comment (isaaclegred) [reply src/PointwiseFunctions/Hydro/Tags.hpp]: I'll rename it

**Review by nilsvu — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
This is great, thank you for doing all this! I've attached a git patch that resolved the compilation issue for me. You can apply using `git am Isaac.patch` in your spectre source directory.

[Isaac.patch](https://github.com/sxs-collaboration/spectre/files/13537729/Isaac.patch)

**Comment by isaaclegred** [comment]
@nilsdeppe this should be ready to go!

**Review by nilsdeppe — APPROVED** [review]
Great, thank you so much for doing this!!
